flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java

```diff
@@ -243,6 +240,13 @@ public void setup(
         stateKeySelector2 = config.getStatePartitioner(1, getUserCodeClassloader());
     }
 
+    private CombinedWatermark setupCombinedWatermark() {
+        CombinedWatermark combinedWatermark = new CombinedWatermark();
+        combinedWatermark.add(new CombinedWatermark.PartialWatermark());
+        combinedWatermark.add(new CombinedWatermark.PartialWatermark());
```

Comment:
shouldn't that depend on the number of inputs? Can't we create that in ctor?

Reply:
I can add an additional ctor. (I was actually thinking about it) However, we still need the `add` method for the `WatermarkOutputMultiplexer`.

Reply:
Hm, haven't thought about the dynamic cases (actually Kafka is not really dynamic but I can see that other sources are). Still the comment stands that I wouldn't expose `PartialWatermark`. 

You could have a similar method to `register` in `CombinedWatermark` that returns the index of the newly added partial. The index is then used in `private final Map<String, Integer> watermarkIndexPerOutputId;` This index can then be used to update the `CombinedWatermark` via index.

This approach makes more sense if you can easily have all side-effects inside the `CombinedWatermark#updateWatermark(int index, Watermark watermark)`. So it avoids the case, where you update the partial and then tell the combined watermark: "hey something changed, please check".

However, the current way may still be valid, if you explicitly want to have the information drift and only update at specific times for performance reasons. I'm currently not in the position to assess that properly.

Reply:
So there are two things here. My comment in this thread referred just to the `add` method, not exposing the `PartialWatermark`. Sure, if the issue was just about the `add` method, I would not expose the `PartialWatermark`.

The second problem is the combination of `PartialWatermark` and `updateCombinedWatermark`. Actually, you wrote the exact reason why we need it:
> However, the current way may still be valid if you explicitly want to have the information drift and only update at specific times for performance reasons. I'm currently not in the position to assess that properly.

That's exactly what happens in the  `WatermarkOutputMultiplexer`. It exposes two access patterns:
* getImmediateOutput
* getDeferredOutput

One immediately updates the combined watermark, and the other one does not. AFAIK the deferred one is used for periodic watermarks. Bear in mind that the `PeriodicWatermark` has the default scope. It is exposed only to the `org.apache.flink.api.common.eventtime` package for it to be used by `WatermarkOutputMultiplexer`.

Reply:
Okay I'd still think that you could also hide that within `#updateWatermark(..., boolean deferred)` but the benefit of encapsulation gets even smaller (if it even still exists). So leave as is.